gadget/update.go

```diff
@@ -904,6 +905,247 @@ func buildNewVolumeToDeviceMapping(old GadgetData, laidOutVols map[string]*LaidO
 	}, nil
 }
 
+func buildVolumeStructureToLocation(old GadgetData,
+	isPreUC20 bool,
+	missingInitialMapping bool,
+	laidOutVols map[string]*LaidOutVolume,
+	volToDeviceMapping map[string]DiskVolumeDeviceTraits,
+) (map[string]map[int]StructureLocation, error) {
+
+	// helper function for handling non-fatal errors on pre-UC20
+	maybeDieOnMappingError := func(err error) (fatal bool) {
+		if missingInitialMapping && isPreUC20 {
+			// this is not a fatal error on pre-UC20
+			logger.Noticef("WARNING: not applying gadget asset updates on main system-boot volume due to error mapping volume to physical disk: %v", err)
+			return false
+		}
+		return true
+	}
+
+	volumeStructureToLocation := make(map[string]map[int]StructureLocation, len(old.Info.Volumes))
+	for volName := range old.Info.Volumes {
+		volumeStructureToLocation[volName] = make(map[int]StructureLocation)
+	}
+
+	// now for each volume, iterate over the structures, putting the
+	// necessary info into the map for that volume as we iterate
+
+	// this loop assumes that none of those things are different between the
+	// new and old volume, which may not be true in the case where an
+	// unsupported structure change is present in the new one, but we check that
+	// situation after we have built the mapping
+	for volName, diskDeviceTraits := range volToDeviceMapping {
+		vol, ok := old.Info.Volumes[volName]
+		if !ok {
+			return nil, fmt.Errorf("internal error: volume %s not present in gadget.yaml but present in traits mapping", volName)
+		}
+
+		laidOutVol := laidOutVols[volName]
+
+		// find the disk associated with this volume using the traits we
+		// measured for this volume
+		validateOpts := &DiskVolumeValidationOptions{
+			// implicit system-data role only allowed on pre UC20 systems
+			AllowImplicitSystemData:     isPreUC20,
+			ExpectedStructureEncryption: diskDeviceTraits.StructureEncryption,
+		}
+
+		disk, err := searchForVolumeWithTraits(laidOutVol, diskDeviceTraits, validateOpts)
+		if err != nil {
+			dieErr := fmt.Errorf("could not map volume %s from gadget.yaml to any physical disk: %v", volName, err)
+			if fatal := maybeDieOnMappingError(dieErr); fatal {
```

Comment:
if you rename the function, then you can also get rid of the `fatal` variables, because the code will be readable even without them.